File: algorithms/2_medium/N/00967_numbers_with_same_consecutive_differences/backtrack.cpp
class Solution {
    static void dfs(const int n, const int diff, vector<int>& res, const int curr) {
        if (n == 0) {
            res.push_back(curr);
            return;
        }

        const int last = curr % 10;
        if (last + diff <= 9) {
            dfs(n - 1, diff, res, (curr * 10) + last + diff);
        }

        if (diff != 0 && last - diff >= 0) {
            dfs(n - 1, diff, res, (curr * 10) + last - diff);
        }
    }

public:
    static vector<int> numsSameConsecDiff(const int n, const int k) {
        assert(n >= 2 && n <= 9 && k >= 0 && k <= 9);
        auto res = vector<int>();

        for (int i = 1; i <= 9; i += 1) {
            dfs(n - 1, k, res, i);
        }

        return res;
    }
};

File: algorithms/2_medium/N/00967_numbers_with_same_consecutive_differences/backtrack.rs
impl Solution {
    fn dfs(n: i32, diff: i32, res: &mut Vec<i32>, curr: i32) {
        if (n == 0) {
            res.push(curr);
            return;
        }

        let last = curr % 10;
        if last + diff <= 9 {
            Self::dfs(n - 1, diff, res, (curr * 10) + last + diff);
        }

        if diff != 0 && last - diff >= 0 {
            Self::dfs(n - 1, diff, res, (curr * 10) + last - diff);
        }
    }

    pub fn nums_same_consec_diff(n: i32, k: i32) -> Vec<i32> {
        assert!(n >= 2 && n <= 9 && k >= 0 && k <= 9);
        let mut res: Vec<i32> = Vec::new();

        for i in 1..=9_i32 {
            Self::dfs(n - 1, k, &mut res, i);
        }

        res
    }
}
